keps/sig-node/4800-cpumanager-split-l3cache/README.md

```diff
@@ -0,0 +1,680 @@
+
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+For `CPUManager` it is a requirement going from `none` to `static` policy cannot be done dynamically because of the `cpu_manager_state file`. The node needs to be drained and the policy checkpoint file (`cpu_manager_state`) need to be removed before restarting Kubelet. This feature specifically relies on the `static` policy being enabled.
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: `CPUManagerAlphaPolicyOptions`
+  - Components depending on the feature gate: `kubelet`
+- [x] Other
+  - Describe the mechanism: Change the `kubelet` configuration to set a `CPUManager` policy of static then setting the policy option of `align-cpus-by-uncorecache`
+  - Will enabling / disabling the feature require downtime of the control
+    plane? No.
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? Yes, a `kubelet` restart is required for changes to take place.
+
+###### Does enabling the feature change any default behavior?
+
+No, to enable this feature, it must be explicitly set in the `CPUManager` static policy and the policy option `align-cpus-by-uncorecache` must be set.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes this feature can be disabled it will just require a restart of `kubelet`. The Kubelet configuration will need to be set with the static policy option and align-cpus-by-uncorecache flag removed.
+
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Feature will be enabled. Proper drain of node and restart of kubelet required. Feature is not intended to be enabled/disabled dynamically, similar to static policy.
+
+###### Are there any tests for feature enablement/disablement?
+
+Option is not enabled dynamically. To enable/disable option, cpu_manager_state must be removed and kubelet must be restarted.
+Unit tests will be implemented to test if the feature is enabled/disabled.
+E2e node serial suite can be use to test the enablement/disablement of the feature since it allows the kubelet to be restarted.
+
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+Kubelet restarts are not expected to impact existing CPU assignments to already running workloads
+
+
+###### What specific metrics should inform a rollback?
+
+N/A
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+N/A
+
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+N/A
+
+
+### Monitoring Requirements
+
+Include CacheID info in podresources API to be able to verify assignment.
```

Comment:
I think there's often the hope that a cluster admin can see easily that this feature is enabled. Do we have a metric the kubelet emits on which cpumanager features are enabled? should we create one? cc @ffromani

Reply:
Note: this requirement scales better if the admin doesn't have to manually inspect each node, but can see in aggregate somewhere

Reply:
This is a good point to bring up @ffromani did mention it before and we were thinking of piggy backing off of his PR he has open

Reply:
I think right now, one metric that can increment would be that we have this option enabled and we did not return an error for the Allocate function

Reply:
yup that'd be perfect!

Reply:
we have few metrics added or being added that can help understand the feature is enabled and working:
- `CPUManagerPinningRequestsTotal` (merged)
- `ContainerAlignedComputeResources` (proposed: https://github.com/kubernetes/kubernetes/pull/127155/commits/dbf0aca80f2f1c2f09cd384f2ce620b8b0ef961b#diff-7a99719139dd480a780b106a988e3175c2a5e72d48cc8399666843a8e631221dR928)

there's not a direct reflection(e.g. node status or metric) which tells if features are enabled (like, say, `cpumanager_aligning` either 0 or 1 acting as flag).
Thinking aloud, I'm not inclined to add those, it seems a bit of shoehorning reporting into metrics, which seems less than ideal to me. I think we're on a good track adding metrics, and more will help, but better visibility of cpumanager (et. al.) behavior and current state is still a open question.

Reply:
Would something like this satisfy this section

Currently metrics are being added to provide more visibility of CPUManager behavior such as `CPUManagerPinningRequestsTotal` and `ContainerAlignedComputeResources` (Proposed). Though a needed metric would be for being able to see if the option is enabled and functioning as intended. 

Also, can you elaborate on how that would be shoehorning into metrics? Wouldn't metrics letting you know what staticpolicy options are enabled a good step forward too?

Reply:
> Would something like this satisfy this section
> 
> Currently metrics are being added to provide more visibility of CPUManager behavior such as `CPUManagerPinningRequestsTotal` and `ContainerAlignedComputeResources` (Proposed). Though a needed metric would be for being able to see if the option is enabled and functioning as intended.

Yes, this should be good enough IMHO. Let's use the external names though, I mistakenly copypasted the internal code name.

E.g. `CPUManagerPinningRequestsTotal` is visible in prometheus as `kubelet_cpu_manager_pinning_requests_total`

> Also, can you elaborate on how that would be shoehorning into metrics? Wouldn't metrics letting you know what staticpolicy options are enabled a good step forward too?

Reply:
don't we have in NFD labeling mechanism to label node based on selected kubelet options? I had impression that it was added some time ago, so people can schedule pods to nodes where specific combination of options enabled.

Reply:
I don't think NFD supports that out-of-the-box I think we would need to follow this customization guide for that. Are we saying we would want to add this into NFD to report back Kubelet options?